### PR TITLE
fix(input, textarea): show error state after touch

### PR DESCRIPTION
**Pull request by liamdebeasi**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue URL: resolves https://github.com/ionic-team/ionic-framework/issues/26939, resolves https://github.com/ionic-team/ionic-framework/issues/21643

There are a couple issues with the form validation highlighting:

1. The valid highlighting was always showing on a valid input, even when the input was blurred.
2. The invalid error text was always showing on an invalid input, even when the input was untouched (touched = the user has focused and subsequently blurred the input)

Form validation highlighting is used to indicate that the user did something right/wrong, so having these states always show can be confusing.

## What is the new behavior?

I updated the behavior as per the following table:

| Validity State | Touched State | Focused State | Validity Highlighting Shown |
| - | - | - | - |
| Valid | Untouched | Blurred | ⚫️ None |
| Valid | Touched | Blurred | ⚫️ None |
| Valid | Untouched | Focused | 🟢 Valid |
| Valid | Touched | Focused | 🟢 Valid |
| Invalid | Untouched | Blurred | ⚫️ None |
| Invalid | Touched | Blurred |  🔴 Invalid |
| Invalid | Untouched | Focused | ⚫️ None |
| Invalid | Touched | Focused | 🔴 Invalid |

Reasoning for invalid state changes: An invalid state indicates the user entered malformed/incorrect data. As a result, having the invalid state show before the user has even interacted with the input can be confusing for the user. We show the input regardless of focus state once it has been touched. This is because users may need to correct data in order to submit a form. Showing the invalid highlight on both focus and blur lets users quickly identify which form controls need fixing.

Reasoning for the valid changes: A valid state indicates that the user entered correct data. As a result, we should give them that positive feedback as soon as their data is valid. We also show the valid state regardless of touched state because some forms may pre-populate data. Users may not necessarily want to change that pre-populated data if it is valid. Additionally, the user does not need to always know that an input is valid which is why we only show it on focus. This behavior also aligns with the behavior found in Ionic 6.

- I found and removed duplicate code in the input and textarea stylesheets
- I updated the input and textarea tests to check the correct valid/invalid conditions
- I added validity state tests for select

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Dev build: 7.0.0-dev.11678373650.1975657a

### CODE REVIEW

File: core/src/components/textarea/textarea.scss [hunk 1]
```diff
@@ -286,17 +286,6 @@
...
-// ----------------------------------------------------------------
-
-:host(.ion-touched.ion-invalid) {
-  --highlight-color: var(--highlight-color-invalid);
-}
-
-:host(.ion-touched.ion-valid) {
-  --highlight-color: var(--highlight-color-valid);
-}
-
```
Comment (liamdebeasi) [inline core/src/components/textarea/textarea.scss:299]: This was duplicate code

File: core/src/components/input/input.scss [hunk 2]
```diff
@@ -398,17 +414,6 @@
...
-// ----------------------------------------------------------------
-
-:host(.ion-touched.ion-invalid) {
-  --highlight-color: var(--highlight-color-invalid);
-}
-
-:host(.ion-touched.ion-valid) {
-  --highlight-color: var(--highlight-color-valid);
-}
-
```
Comment (liamdebeasi) [inline core/src/components/input/input.scss:411]: This was duplicate code

**Review by sean-perkins — APPROVED** [review]
Changes look good. 

The test templates (e.g. test/bottom-content/index.html) are a little confusing to reason about right now, unless you are aware of the extra classes that need to be appended to the host to make the error slot visible and/or the styling. Should we add a minimal implementation here with a button that adds the required classes to help with manually testing?

**Comment by liamdebeasi** [comment]
Good catch! I updated those templates so `ion-touched` is added and the error styles show up.

**Pull request merged**